\chapter*{Introduction}
\addcontentsline{toc}{chapter}{Introduction}
\phantomsection{}
\markboth{Introduction}{Introduction}

Advances in new DNA sequencing technologies lead to rapidly increasing number of available sequences from different organisms, and along with that increases the for efficient processing of such data.

The output from the sequencing machines is a huge number of short reads which are random substrings of the original sequence. In addition the reads may contain errors.
These reads are then assembled together, but often it is not possible to fully reconstruct the original DNA sequence.

Therefore it is important to develop algorithms for analysis the sequencing reads, which can give us information about the DNA sequence without a need for a full sequence assembly.
In this work, we are interested in two kinds of information: the size of the genome and characteristics of the repetitive sequences in the genome.

This work is organized in five chapters.
In the first chapter, we introduce necessary biological background. We define a genome, briefly explain a DNA sequencing and sequence assembly process.
In the second chapter, we discuss several existing approaches for $k$-mer counting, which is a crucial method in many DNA sequencing data analyses.
In the third chapter, we focus on algorithms for genome size and coverage estimation. We include our approach to this problem, used in CovEst~\cite{covest}, which we presented at the SPIRE 2015 conference.
In the fourth chapter, we summarize several approaches for the repetitive sequence search under various conditions. We also describe a gene family evolution model, which provides a null hypothesis for gene family evolution.
In the last chapter, we set a goals for our PhD project. We want to focus on improvements of our genome size estimation tool and to develop new methods for gene family size estimation, incorporate it to the gene family evolution model and develop a repeat evolution model.
