\chapter*{Introduction}
\addcontentsline{toc}{chapter}{Introduction}
\phantomsection
\markboth{Introduction}{Introduction}

New sequencing technologies are producing an increasing number of sequences from different organisms, and along with that the need for understanding such data.

The output from the sequencing machines, is a huge number of short reads, which are random substrings of the original data, which may contain some errors. To obtain a full DNA sequence, the DNA assembling is needed. Many times, the DNA assembling is not possible, due to several issues of the complex DNA assembly process.

Therefore it is important to develop algorithms for analysis the sequencing reads, which can give us a lot of information about the DNA sequence without a need for a full DNA assembly.
In this work, we are interested in two kinds of information: the size of the genome and characteristic of the repetitive sequences in the genome.

This work is organized in five chapters.
In the first chapter, we introduce some biological background for this work. We define a genome, briefly explain a DNA sequencing and DNA assembling process.
In the second chapter we introduce several approaches for \kmer\ counting, which is a crucial method in most DNA sequencing data analyses.
In the third chapter, we focus on algorithms for genome size and coverage estimation algorithms. We include our approach to this problem, CovEst\cite{covest}, which we presented on SPIRE 2015 conference.
In the fourth chapter, we summarize several approaches to the repetitive sequence finding under various conditions. We also introduce a gene family evolution model, which provides a null hypothesis for gene family evolution.
In the last chapter, we set a goals for our PhD project. We want to focus on improvements of our genome size estimation tool and we want to develop new methods for gene family size estimation, incorporate it to the gene family evolution model and develop a repeat evolution model.
