\chapter{Introduction}
%
% New sequencing technologies are producing more and more biological data,
% including genomic sequences of many species. Therefore it is important to
% develop tools for automated analysis of such data. In this thesis we focus on
% computational methods for sequence annotation and sequence alignment.  In the
% sequence annotation problem, we want to label parts of the sequences according
% to their function, or meaning. We call
% such a labeling an \firstUseOf{annotation}. For example,  we
% can label each symbol of a genomic sequence base on whether it is part of a
% gene or not as in the following example ($g$ is a label representing genes and $n$ is a label for non-gene parts).
% \begin{verbatim}
% Sequence:     ACGGTGCGTTAGCTGCTCTGATGTCTTCGATCTAGCTAGT
% Annotation:   nnnnnnnngggggggggggggggggggnnnnnnnnnnngg
% \end{verbatim}
% The \firstUseOf{sequence alignment} is a data structure
% that characterizes similarity or shared origin of two or more sequences.
% We insert gap symbols (dashes) so that corresponding parts of the sequence are in the same column as in the following
% example.
% \begin{verbatim}
% Sequence X:   CTGCTAGCTACGT--GTGT
% Sequence Y:   ---------ACGTGGAT--
% \end{verbatim}
% Both annotation and alignment are fundamental bioinformatics problems.
% The first stages of analysis of newly
% sequences genomes typically include aligning it with the genomes of related
% species (that is already sequenced), and searching for known structures (like genes) inside
% new genomes.
% Many subsequent methods for analysing genomes rely on sequence
% annotation and alignment. To avoid artefacts in the results of
% these downstream methods, there is need to develop algorithms for producing sequence
% annotation and alignment with as low error rate as possible.
% Tools for both
% sequence annotation and alignment are often based on hidden Markov models (HMMs)
% \cite{Durbin1998,Alexanderson2004,Brejova2005, FEAST2011,Krogh2001,Majoros2005,
% Meyer2002,Nanasi2010,Pairagon2009, Schultz2006,Kovac2012,Pachter2002,
% Liu2010,Brown2010, Lunter2008}.  In this thesis we propose new techniques for
% use of HMMs  in these domains and also give proofs of NP-hardness for several related problems.
%
% We work with generative probabilistic models, \abbreviation{hidden Markov
% Models}{HMM} and their variants. In general, an HMM is a state machine that
% generates a sequence (string) along with a sequence of states (called state path).
% Since an HMM is a probabilistic model,
% it also defines the probability of sequences and state paths. The state
% path contains information about the structure of the generated sequence. In
% practice we are often given the generated sequence and the state path is hidden. The
% goal of the decoding algorithm is to reverse the generation process and obtain
% the original state path or at least its approximation.
%
% When using HMMs for annotation of biological sequences, we construct the
% HMM so  that the structure of the states corresponds to the biological features we
% are interested in.
% Each feature can be encoded in one or several states.  Then we assume that the genomic sequence of interest was generated by our model and use a decoding
% algorithm to obtain a state path which is as close as possible to the true
% state path. The Viterbi algorithm
% \cite{Durbin1998} is traditionally used for decoding, but other optimization criteria can be used to obtain more
% accurate results \cite{Brown2010, Gross2007, Nanasi2010, Truszkowski2011}. Note that
% accuracy measures may depend on the application domain.  In Chapter
% \ref{CHAPTER:TWOSTAGE} we study a special type of decoding algorithms: two-stage
% algorithms. In the first stage, the algorithm infers important aspects of
% the annotation and in the second stage it fills remaining details in a way
% consistent with the first-stage results.  We show that  two-stage
% algorithms can improve the accuracy of decoding (as far we know, such
% algorithms were previously used only for reducing the running time). We also
% study the computational complexity of several decoding criteria appropriate for
% the first stage and we show NP-hardness results for obtaining the optimal
% annotations using these criteria. Namely we study the most probable footprint
% problem, the most probable set problem and the most probable restriction
% problem.
%
% In sequence alignment, the goal is to search for
% corresponding  parts of the sequences and arrange them into same position in the
% alignment.  To choose the biologically correct alignment, we usually optimize some
% scoring scheme. We will consider scoring schemes which are  defined using
% \abbreviation{pair hidden Markov models}{pHMM}. A pHMM generates pairs of
% sequences along with their alignment (an alignment is defined by the state
% path).  This model is an extension of HMM.
%
%
% In the fourth chapter we propose a tractable method for aligning sequences with
% tandem repeats. A tandem repeat consists of consecutive copies (not exact) of a
% certain motif (short genomic sequence). Tandem repeats cause problems with
% sequence alignments because it is hard to distinguish between individual copies of the motif.
% We extend a traditional pHMM model for sequence alignment by additional states
% modeling tandem repeats. We also propose new decoding algorithms tailored to
% this model. We show on simulated data that our new model and decoding methods
% decrease the error rate, and with a particular increase of accuracy near the border of tandem repeats.
%
\section{Biological Background}\label{SECTION:BIOINTRO}

\subsection{Genome}

Each life form is build from at least one cell. One of the most important part of the cell is \emph{DNA}, which is located in the cell core.
DNA (Deoxyribonucleic acid) carries the genetic information used for development, functioning and reproduction of the cell or an organism.
DNA is composed of two complementary strains which form a double helix structure. Each strain consist of four nucleotides: cytosine (C), guanine (G), adenine (A), or thymine (T).
The nucleotides are complementary, i.e.\ if in one strain there is A in the other is T, and if in one is G in the other is C, see Figure~\ref{}\todo{figure}.
DNA molecules are organized in chromosomes.
The \emph{genome} is a set of all chromosomes in the cell, it includes both the coding and the non-coding parts of DNA.\@
The coding part of the genome is called gene. Each gene can be translated into a protein sequence which it encodes.

\subsection{DNA Sequencing and Assembly}

Obtaining a DNA sequence or its part from the cell is called \emph{DNA sequencing}.
DNA sequencing connects experimental methods of obtaining the sequence from the cell core and the computational analysis of the sequencing data called sequence assembling.

The first sequenced genome was a simple RNA virus in 1976. The 3.2Gb long human genome was first sequenced in 2001 and it took 13 years and cost about 3 billions dollars. Nowadays it is possible to sequence such large genome for much lower cost (several tens thousands dollars).

The traditional sequencing method, called \emph{Sanger sequencing}, was developed by Frederick Sanger in 1977\cite{sanger1977dna}. It was also used in first human genome sequencing.
In the Sanger sequencing, two strand DNA is first split into single strand DNA.\@ Such single strand DNA can be used to create its copies, if particular enzymes and sufficient amount of nucleotides is available. The key role in Sanger sequencing form modified nucleotides. The nucleotides are modified in two ways: they are colored by florescent coloring and modified such the DNA copy process cannot continue after the nucleotide. This yields in bunch of DNA segments of various lengths, which are basically the sequenced DNA prefixes interrupted by the colored nucleotide in its last position. Every position is represented by multiple segments. The segments are then sorted using electrophoresis and the colors are read from shortest to longest segment forming a DNA sequence.

Only short sequences, up to 500--1000 bases, can be read this way. Therefore the original DNA has to be split into many overlapping short sequences prior the sequencing.
Therefore instead of one long sequence per chromosome, we get many short sequences which have to be assembled together to obtain the whole sequence.

Since 2005, there have been commercially available sequencing machines of second generation (or new generation sequencing machines). There are multiple different technologies for NGS, but they all share some properties.
They produce a huge amount of reads from the same sample, and are faster and cheaper. They also produce much shorter reads (35--400 bases) then Sanger sequencing, which makes a DNA assembly even harder task.

\subsubsection{DNA Assembly}

\todo{ked to zosekvenujeme treba to poskladat, spomenut priblizne princip, par toolov a problemy ake to ma --- coverage treba vysoky, problemy s repeatmi}

\end{itemize}

\todo{chceme aj celu minimovku nejak zhrnut a dat motivaciu k tym repeatom}
